Bluetooth: Add support for Path Loss Monitoring feature

sean-madigan opened the pull request:
This commit adds host support for the Path Loss Monitoring feature see Bluetooth Core specification, Version 5.4, Vol 6, Part B, Section 4.6.32.

Limited logic is required, just adding a wrapper around the HCI command for ease of use.

Add new zone - BT_CONN_LE_PATH_LOSS_UNAVAILABLE, to convert 0xFF path loss to a useful zone.

Add new Kconfigs and functionality to the bt shell.

hermabe commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,32 @@ struct bt_conn_le_tx_power_report {
 	int8_t delta;
 };
 
+enum bt_conn_le_path_loss_zone {
```
There should probably be some doxygen comments on the enum and the values.

hermabe commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,32 @@ struct bt_conn_le_tx_power_report {
+	BT_CONN_LE_PATH_LOSS_UNAVAILABLE = 3,
+};
+
+/** LE Path Loss Monitoring Threshold Change Report Structure*/
```
`@brief` might be needed for the comments to be rendered in the docs

hermabe commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,32 @@ struct bt_conn_le_tx_power_report {
+	uint8_t path_loss_dbm;
+};
+
+/** LE Path Loss Monitoring Parameters Structure */
```
Maybe reference the HCI spec for the definition of these parameters?

hermabe commented on subsys/bluetooth/host/conn.c:
```diff
@@ -2884,6 +2884,66 @@ int bt_conn_le_get_tx_power_level(struct bt_conn *conn,
+	cp->high_hysteresis = params->high_hysteresis;
+	cp->low_threshold = params->low_threshold;
+	cp->low_hysteresis = params->low_hysteresis;
+	cp->min_time_spent = params->min_time_spent;
```
This is 16 bits so needs endianness conversion as well.
```suggestion
	cp->min_time_spent = sys_cpu_to_le16(params->min_time_spent);
```

hermabe commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2929,6 +2958,51 @@ static int cmd_set_power_report_enable(const struct shell *sh, size_t argc, char
+
+static int cmd_set_path_loss_reporting_enable(const struct shell *sh, size_t argc, char *argv[])
+{
+	if (argc != 2) {
```
Are these `argc` checks needed? I thought the `SHELL_CMD_ARG` macro defined the number of required and optional arguments.

hermabe reviewed (commented):
Looks good, just some minor comments

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,46 @@ struct bt_conn_le_tx_power_report {
+ *  @note BT_CONN_LE_PATH_LOSS_UNAVAILABLE has been added to notify when path loss becomes
+ *        unavailable.
+ */
+enum bt_conn_le_path_loss_zone {
+	/* Low path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ENTERED_ZONE_LOW = 0,
```
Consider using the same prefix for the values as the name of the enum. 
This is a fairly common practice, even if we don't have specific coding guidelines for it:
```suggestion
enum bt_conn_le_path_loss_zone {
	/* Low path loss zone entered. */
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_LOW = 0,
```

hermabe replied:
Another one for https://github.com/zephyrproject-rtos/zephyr/issues/55281 I guess

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,46 @@ struct bt_conn_le_tx_power_report {
+/** @brief LE Path Loss Monitoring Threshold Change Report Structure. */
+struct bt_conn_le_path_loss_threshold_report {
+
+	/** Path Loss zone as documented in Core Spec. Version 5.4 Vol.4, Part E, 7.7.65.32. */
+	enum bt_conn_le_path_loss_zone zone_entered;
```
```suggestion
	/** Path Loss zone as documented in Core Spec. Version 5.4 Vol.4, Part E, 7.7.65.32. */
	enum bt_conn_le_path_loss_zone zone;
```

Since the value can also be `BT_CONN_LE_PATH_LOSS_UNAVAILABLE`, using `entered` in the field name seems incorrect (or at least unnecessary).

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,46 @@ struct bt_conn_le_tx_power_report {
+/** @brief LE Path Loss Monitoring Parameters Structure as defined in Core Spec. Version 5.4
+ *         Vol.4, Part E, 7.8.119 LE Set Path Loss Reporting Parameters command.
+ */
+struct bt_conn_le_path_loss_reporting_parameters {
+	uint8_t high_threshold;
+	uint8_t high_hysteresis;
+	uint8_t low_threshold;
+	uint8_t low_hysteresis;
+	uint16_t min_time_spent;
+};
```
Similar to the enum above, all fields in structs should be documented

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -601,6 +641,31 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
+ *
+ *  @return Zero on success or (negative) error code on failure.
+ */
+int bt_conn_le_set_path_loss_monitoring_parameters(struct bt_conn *conn,
+				const struct bt_conn_le_path_loss_reporting_parameters *param);
```
Indentation looks weird here, but may always look weird due to the length of the struct name.

Consider shortening some of the names to e.g.
```suggestion
int bt_conn_le_set_path_loss_mon_param(struct bt_conn *conn,
				const struct bt_conn_le_path_loss_param *param);
```

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2929,6 +2958,43 @@ static int cmd_set_power_report_enable(const struct shell *sh, size_t argc, char
+		return -ENOEXEC;
+	}
+	const struct bt_conn_le_path_loss_reporting_parameters params = {
+		.high_threshold = strtoul(argv[1], NULL, 10),
+		.high_hysteresis = strtoul(argv[2], NULL, 10),
+		.low_threshold = strtoul(argv[3], NULL, 10),
+		.low_hysteresis = strtoul(argv[4], NULL, 10),
+		.min_time_spent = strtoul(argv[5], NULL, 10),
```
Suggest to use the `shell_stroul` functions and verify the input before calling `bt_conn_le_set_path_loss_monitoring_parameters`

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -189,6 +189,24 @@ static const char *enabled2str(bool enabled)
 #endif /* CONFIG_BT_TRANSMIT_POWER_CONTROL */
 
+#if defined(CONFIG_BT_PATH_LOSS_MONITORING)
+static const char *plm_report_zone2str(enum bt_conn_le_path_loss_zone zone_entered)
```
In Bluetooth in Zephyr it seems more common to either do 
```suggestion
static const char *plm_report_zone_str(enum bt_conn_le_path_loss_zone zone_entered)
```
or
```suggestion
static const char *plm_report_zone_to_str(enum bt_conn_le_path_loss_zone zone_entered)
```

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,53 @@ struct bt_conn_le_tx_power_report {
+ */
+enum bt_conn_le_path_loss_zone {
+	/* Low path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_LOW = 0,
+	/* Middle path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_MIDDLE = 1,
+	/* High path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_HIGH = 2,
+	/* Path loss has become unavailable. */
+	BT_CONN_LE_PATH_LOSS_ZONE_UNAVAILABLE = 3,
```
It is OK to keep the values if you prefer, but otherwise it can also just be
```suggestion
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_LOW,
	/* Middle path loss zone entered. */
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_MIDDLE,
	/* High path loss zone entered. */
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_HIGH,
	/* Path loss has become unavailable. */
	BT_CONN_LE_PATH_LOSS_ZONE_UNAVAILABLE,
```

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,53 @@ struct bt_conn_le_tx_power_report {
+	/** Path Loss zone as documented in Core Spec. Version 5.4 Vol.4, Part E, 7.7.65.32. */
+	enum bt_conn_le_path_loss_zone zone;
+
+	/**Current path loss in dBm */
+	uint8_t path_loss_dbm;
```
```suggestion
	/**Current path loss in dB */
	uint8_t path_loss_db;
```
![image](https://github.com/zephyrproject-rtos/zephyr/assets/2520413/64addbf4-0bce-4969-899e-60a473130559)

If we want to keep `_db` as part of the field consider if we should add it to the fields of `bt_conn_le_path_loss_reporting_param` as well

sean-madigan replied:
Removed dB, hopefully more consistent now?

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,53 @@ struct bt_conn_le_tx_power_report {
+	/** Path Loss zone as documented in Core Spec. Version 5.4 Vol.4, Part E, 7.7.65.32. */
+	enum bt_conn_le_path_loss_zone zone;
+
+	/**Current path loss in dBm */
```
```suggestion
	/** Current path loss in dBm */
```

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,53 @@ struct bt_conn_le_tx_power_report {
+ *         Vol.4, Part E, 7.8.119 LE Set Path Loss Reporting Parameters command.
+ */
+struct bt_conn_le_path_loss_reporting_param {
+	/* High threshold for the path loss (dB). */
+	uint8_t high_threshold;
+	/* Hysteresis value for the high threshold (dB). */
+	uint8_t high_hysteresis;
+	/* Low threshold for the path loss (dB). */
+	uint8_t low_threshold;
+	/* Hysteresis value for the low threshold (dB). */
+	uint8_t low_hysteresis;
+	/* Minimum time in number of connection events to be observed once the
+	 * path loss crosses the threshold before an event is generated.
+	 */
```
You need to use either `/**` or `/** @brief` for doxygen to pick it up
```suggestion
	/** High threshold for the path loss (dB). */
	uint8_t high_threshold;
	/** Hysteresis value for the high threshold (dB). */
	uint8_t high_hysteresis;
	/** Low threshold for the path loss (dB). */
	uint8_t low_threshold;
	/** Hysteresis value for the low threshold (dB). */
	uint8_t low_hysteresis;
	/** Minimum time in number of connection events to be observed once the
	 *  path loss crosses the threshold before an event is generated.
	 */
```

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -515,6 +515,53 @@ struct bt_conn_le_tx_power_report {
+ *        unavailable.
+ */
+enum bt_conn_le_path_loss_zone {
+	/* Low path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_LOW = 0,
+	/* Middle path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_MIDDLE = 1,
+	/* High path loss zone entered. */
+	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_HIGH = 2,
+	/* Path loss has become unavailable. */
```
```suggestion
	/** Low path loss zone entered. */
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_LOW = 0,
	/** Middle path loss zone entered. */
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_MIDDLE = 1,
	/** High path loss zone entered. */
	BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_HIGH = 2,
	/** Path loss has become unavailable. */
```

Thalley commented on include/zephyr/bluetooth/conn.h:
```diff
@@ -601,6 +648,30 @@ int bt_conn_le_set_tx_power_report_enable(struct bt_conn *conn,
+ *
+ *  @return Zero on success or (negative) error code on failure.
+ */
+int bt_conn_le_set_path_loss_mon_param(struct bt_conn *conn,
+				       const struct bt_conn_le_path_loss_reporting_param *param);
+
+/** @brief Enable or Disable Path Loss Monitoring.
+ *
+ * Enable or disable Path Loss Monitoring, which will decide whether Path Loss Threshold events
+ * are sent from the controller to the host.
+ *
+ * @param conn   Connection Object.
+ * @param enable Enable/disable path loss reporting.
+ *
+ * @return Zero on success or (negative) error code on failure.
+ */
+int bt_conn_le_set_path_loss_mon_enable(struct bt_conn *conn, bool enable);
+
```
Since the implementations are guarded by `CONFIG_BT_PATH_LOSS_MONITORING`, we should add a note about that here too

Thalley reviewed (changes requested):
A few more comments

Thalley reviewed (dismissed):
LGTM

I assume that the command has been tested with a non-Zephyr controller, right?

sean-madigan commented:
> LGTM
> 
> I assume that the command has been tested with a non-Zephyr controller, right?

Yes, tested most recent changes with SDC using shell

alwa-nordic commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2929,6 +2958,55 @@ static int cmd_set_power_report_enable(const struct shell *sh, size_t argc, char
+		shell_error(sh, "Conn handle error, at least one connection is required.");
+		return -ENOEXEC;
+	}
+	if (*argv[1] == '0') {
```
This is equivalent to the following:

```suggestion
	if (argv[1][0] == '0') {
```

And this has some funny properties, like the empty string is parsed as true, and the string "01" as false.

Wouldn't it be better to use `shell_strtobool`?

sean-madigan replied:
Yes, good idea - updated.

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2929,6 +2958,62 @@ static int cmd_set_power_report_enable(const struct shell *sh, size_t argc, char
+		return SHELL_CMD_HELP_PRINTED;
+	}
+
+	if (enable) {
+		return bt_conn_le_set_path_loss_mon_enable(default_conn, true);
+	}
+
+	return bt_conn_le_set_path_loss_mon_enable(default_conn, false);
```
```suggestion
	return bt_conn_le_set_path_loss_mon_enable(default_conn, enable);
```

Ideally, as per the shell API, the command handler should only return one of 4 values:
```c
/**
 * @brief Shell command handler prototype.
 *
 * @param sh Shell instance.
 * @param argc  Arguments count.
 * @param argv  Arguments.
 *
 * @retval 0 Successful command execution.
 * @retval 1 Help printed and command not executed.
 * @retval -EINVAL Argument validation failed.
 * @retval -ENOEXEC Command not executed.
 */
typedef int (*shell_cmd_handler)(const struct shell *sh,
				 size_t argc, char **argv);
```

So ideally you'd check if `bt_conn_le_set_path_loss_mon_enable` returns either `0` and then return `0`, or if it returns a non-zero value then return `-ENOEXEC`.

sean-madigan replied:
Makes sense, updated

Thalley reviewed (dismissed):
Minor comment, nothing that really needs to be fixed

Thalley commented on subsys/bluetooth/shell/bt.c:
```diff
@@ -2929,6 +2958,76 @@ static int cmd_set_power_report_enable(const struct shell *sh, size_t argc, char
+		return SHELL_CMD_HELP_PRINTED;
+	}
+
+	if (enable) {
+		err = bt_conn_le_set_path_loss_mon_enable(default_conn, true);
+	} else {
+		err = bt_conn_le_set_path_loss_mon_enable(default_conn, false);
+	}
```
```suggestion
	err = bt_conn_le_set_path_loss_mon_enable(default_conn, enable);
```

Thalley reviewed (dismissed):
Again, minor comment :) But doesn't need to be fixed

alwa-nordic commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2502,6 +2502,35 @@ void bt_hci_le_transmit_power_report(struct net_buf *buf)
+	struct bt_hci_evt_le_path_loss_threshold *evt;
+	struct bt_conn_le_path_loss_threshold_report report;
+	struct bt_conn *conn;
+
```
I think we should code defensively against the controller. 

```suggestion

	if (buf->len < sizeof(*evt)) {
		LOG_ERR("Too short bt_hci_evt_le_path_loss_threshold");
		return;
	}
```

We should also maybe check that `evt->zone_entered` and `evt->current_path_loss` is one of the expected values before we pass it on to the application. @jhedberg What do you think about this? The application should also be defensively coded, so it may be fair to just pass on the value as is.

sean-madigan replied:
Is this doing the same thing in the generic event handling code? https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/host/hci_core.c#L158

alwa-nordic replied:
@sean-madigan You are right. Please ignore that part of the suggestion.

sean-madigan replied:
Happy to add a check that `evt->zone_entered` is a sane value. I think any value of `evt->current_path_loss` _could_ be valid

Thalley replied:
I tend to agree with @alwa-nordic that we should be defensive, even against the controller. 

This is something that we today do quite poorly, as we generally tend to trust the controller (which we kind of need to do in order to work in the first place). 

In any case, I'd be for checking that the value is in a valid range

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2502,6 +2502,41 @@ void bt_hci_le_transmit_power_report(struct net_buf *buf)
+	evt = net_buf_pull_mem(buf, sizeof(*evt));
+
+	if (evt->zone_entered > BT_CONN_LE_PATH_LOSS_ZONE_ENTERED_HIGH) {
+		LOG_ERR("Invalid zone in bt_hci_evt_le_path_loss_threshold");
```
```suggestion
		LOG_ERR("Invalid zone %u in bt_hci_evt_le_path_loss_threshold", evt->zone_entered);
```

For debugging purposes it's a good idea to also log the invalid value :)

Thalley commented on subsys/bluetooth/host/hci_core.c:
```diff
@@ -2502,6 +2502,41 @@ void bt_hci_le_transmit_power_report(struct net_buf *buf)
+		report.zone = BT_CONN_LE_PATH_LOSS_ZONE_UNAVAILABLE;
+		report.path_loss = BT_HCI_LE_PATH_LOSS_UNAVAILABLE;
+	} else {
+		report.zone = evt->zone_entered;
```
Just noticed that this assume that the HCI zone definitions are equal to the `enum bt_conn_le_path_loss_zone`, which can be dangerous to assume. 

This can be ensured via BUILD_ASSERTS or by adding a converting function from HCI to Zephyr values similar to 
```c
uint8_t bt_get_phy(uint8_t hci_phy)
{
	switch (hci_phy) {
	case BT_HCI_LE_PHY_1M:
		return BT_GAP_LE_PHY_1M;
	case BT_HCI_LE_PHY_2M:
		return BT_GAP_LE_PHY_2M;
	case BT_HCI_LE_PHY_CODED:
		return BT_GAP_LE_PHY_CODED;
	default:
		return 0;
	}
}
```

sean-madigan replied:
Added build asserts

sean-madigan committed:
Bluetooth: Add support for Path Loss Monitoring feature

This commit adds host support for the Path Loss Monitoring
feature see Bluetooth Core specification, Version 5.4,
Vol 6, Part B, Section 4.6.32.

Limited logic is required, just adding a wrapper around the
HCI command and callback for HCI event.

Add new zone - BT_CONN_LE_PATH_LOSS_ZONE_UNAVAILABLE, to
convert 0xFF path loss to a useful zone.

Add new Kconfigs and functionality to the bt shell.

Signed-off-by: Sean Madigan <sean.madigan@nordicsemi.no>

sean-madigan commented:
@jhedberg are you able to review?
